readme.txt
Rodent, a UCI chess playing engine derived from Sungorus 1.4
Copyright (C) 2009-2011 Pablo Vazquez (Sungorus author)
Copyright (C) 2011-2012 Pawel Koziol
http://www.pkoziol.cal24.pl/rodent/rodent.htm
https://github.com/nescitus/rodent_code

I. GPL License

Rodent is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published 
by the Free Software Foundation, either version 3 of the License, 
or (at your option) any later version.

Rodent is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

II. Acknowledgments

Rodent is an open source chess program derived from Sungorus 1.4 
by Pablo Vazquez. Seeing the simplicity of his code I instantly
fell in love with it, started adding chess knowledge and advanced
search algorithms. Hopefully code quality didn't deteriorate too
much because of that attempt at learning by emulation.

The second most important source is excellent document called
"Toga LOG user manual" by josephd, which is accessible from
http://members.aon.at/josefd/Toga%20LOG.html . It describes 
evaluation function of Fruit/Toga in a natural language. First
draft of Rodent evaluation function relied heavily on this
document. Currently at least king safety code is different, 
but for example weak pawn evaluation remains basically the same.

Open source programming makes most sense as a collaborative effort.
Rodent would be clearly weaker without the following contributions:

*Kestutis Gasaitis* has made a thorough inspection of Rodent's code,
pointing out many bugs and suggesting several improvements. See
changelog and github tickets for details.

*Dann Corbit* supplied fast population count functions 
making use of advanced 64-bit instructions and first bit intrinsics
(used since version 0.11 onwards)

*Denis Mendoza* and *Graham Banks* independently pointed out 
an important time management bug in version 0.12, 
affecting repeated time controls.

*Dann Corbit* and *Denis Mendoza* supplied me with 64-bit compiles.

III. Project goals:

1. Creating a didactic bitboard engine of reasonable strength, 
   using object-oriented design, modern search techniques and
   relatively small codebase. There are many superb open source
   engines, like Fruit, Crafty, Stockfish, to name but a few,
   and Rodent is unlikely to compete with them on equal footing
   for some time to come. One thing it can give, however, is 
   a search function that can be read at one sitting (thank You,
   Pablo Vazquez, it would not be possible without Your work!)
   
2. Providing a simple framework for testing new ideas and showcasing
   them (I'm planning a short paper on a "sliding LMR" concept) 
   
3. Providing an engine that is fun to play against, with adjustable
   strength and different personalities.
   
IV. The most important changes (comparing with Sungorus 1.4) include:

- making most of the code object-oriented
- enabling fractional extensions and reductions
- rewriting piece/square code, so that we can use asymmetric pst tables 
  and interpolate between midgame and endgame scores
- adding futility pruning
- adding "sliding LMR" (see section on search)
- adding eval pruning (inspired by DiscoCheck)  
- adding many evaluation features and weights from Toga LOG user manual  
- adding logarithmic king safety function  
- some speed optimizations, including pawn hash table and specialized
  PopCnt functions (many thanks to Dann Corbit for his patch making use
  of advanced 64-bit instructions)
- opening book in a silly proprietary format
- position learning
- weaker levels of play

V. SEARCH (featuring "sliding LMR" proof of concept)

- fail soft alpha beta with principal variation search (from Sungorus)
- transposition table (from Sungorus)
- null move with variable reduction depth (modified)
  and with threat detection in order to sort evasion higher (modified)
- futility pruning (added)
- "sliding" late move reduction (added)
- late move pruning (added)
- eval pruning (a.k.a. static null move)
- internal iterative deepening in pv nodes

Rodent uses PVS search with null move and transposition table inherited 
from Sungorus. It has been enhanced in several standard ways - by adding 
futility pruning, internal iterative deepening in pv nodes, check extension,
Stockfish-like null move pruning and using capture threats uncovered by 
null move for better move sorting.

The only non-standard solution is what I call "sliding Late Move Reduction".
Since Rodent from the very beginning uses fractional plies, it made sense
to check whether smoothly scaling reduction can help. And it certainly did.

It works as follows. Instead of a fixed one ply reduction, Rodent uses
something like:

QUARTER_PLY * ( (moves_tried / 5)+3 )

where moves_tried counts moves accepted by search as legal. Thus the first 
reduction occurs much later under normal scheme, but as the search proceeds, 
the effect of accumulated quarter ply reductions compensates for it, without 
hurting the playing strength. This kind of reduction has a chance of being 
much less detrimental to the tactical ability of the engine. After all, 
it is geared towards reducing more deeply when the engine encounters a series 
of moves that have low probability of changing the node value (move ordering, 
especially using history heuristic, is a measure of that probability).

Imagine a tactical shot requiring two quiet moves, the rest being forcing 
and therefore exempt from reduction. Both are sorted around move 20. Under 
normal scheme both would get reduced, the new one reduces just one. Instead, 
it trims the lines where many moves are sorted down the list. Finding really 
deep combintations, full of quiet moves, is equally difficult under both 
schemes. However, probability of such a combination is  

P(one_move_cuts) * P(next_move_cuts) ... * P(last_required_move_cuts) 

IIRC (I've never been much into maths, much less into the English math 
terminology) the right word for this kind of function is "exponential". 
If we manage to keep the growth of a reduction factor just a bit slower, 
we'll have a beautifully scalable reduction algorithm.

VI. EVALUATION

- Fruit-like piece/square tables
- mobility (excluding squares attacked by pawns in case of minor pieces)
- Fruit-like weak pawns eval
- passed pawns eval
- outposts (B, N, even R), further bonus if defended by a pawn
- logaritmic, and therefore dynamic King safety evaluation
  (with bigger bonus for the first attack by a piece) 
- possibility to give check and safe queen contact checks
- hanging pieces
- minor pieces attacked/defended by a pawn (overlaps with hanging pieces)
- bad bishops (using bitmask of obstructing pawns)
- (half)open files for rooks with additional bonus if they target enemy king
- rook on 7th rank cutting off enemy king or attacking pawns

Creating version 0.13 I looked for more aggressive King safety function,
using Rebel-like additive table approach. I also wanted my function to 
ne initialized at startup using some relatively simple mathematical formula.
After some experiments with logistic function, I settled for an addition
of logarithmic and linear function. I'd like to express my gratitude
to Agnieszka Ziomek for discussing these mathematical matters with a noob
like me.

Mobility evaluation is sometimes inexact due to the concept of "transparent 
pieces". For example, own queen is considered transparent for the bishop. 
This might cause occasional error in evaluating whether a bishop can check 
enemy king, or force the engine to rely on search in order to avoid certain
blockages. However, tests has shown that this solution is better than more
exact mobility calculation.

VII. OPENING BOOK

Since version 0.14 Rodent uses its own opening book, encoded in a rather
inefficient manner. At startup, it reads two files: one of guide books 
written in a human readable format, stored in books folder, and bigbook.wtf 
using Rodent's own ugly format (hence the file extension). 

Rodent first looks for moves from the guide book, and if none is found, 
it falls back to big main book.

guide books encodes one opening in each line, and its format looks like that:

e2e4 a7a6? d2d4 b7b5 a2a4 c8b7 b1d2
e2e4 d7d5!! e4d5 d8d5 b1c3 d8d6!

It allows users to write down their own book preferences if they for example 
want to practice Ruy Lopez or Pirc Defense or force Rodent to play unusual 
openings. Moves may be followed by punctuation marks: ? (frequency -= 100),
! (frequency += 100), ?? (frequency -= 5000), !! (frequency += 5000). 
Additionally, a move with "xx" appended will be deleted from the main book
permanently.

A really desperate user might even create main opening book. All he needs
is any of files called feed.txt, feedwhite.txt and feedblack.txt placed 
in the program directory, Rodent opened in the console mode and manually 
typed command "feedbook". Program first reads bigbook.wtf (if found), then 
supplements it with the lines froom feed files, then chokes itself on shaker 
sort algorithm while displaying number of its iterations, and finally saves 
a file called newbook.wtf. At this stage user MUST RENAME newbook.wtf 
to bigbook.wtf. This is a precaution allowing to save old book in case that
user considers feed files corrupt.

Calling "feedbook" with a numeric parameter causes Rodent to verify new
book moves with a search to the depth supplied by the parameter. It takes
insanely long, though.

Rodent contains small utility reformatting book files from something like:

e2e4e7e6d2d4d7d5b1c3f8b4

to version with spaces between the moves. In order to do this, one needs
to copy "dense" opening book into the file "dense.txt", open Rodent in the
console mode and type command "split". Result will be saved in "loose.txt".

Another additional command is "bookdoctor". This command causes Rodent to
investigate its main book file and stop at first point where it finds
either an infrequent move or missing transposition into book line. In both
cases it would be prudent to decide whether a move should be upgraded
or marked as unplayable.

VIII. WEAKENING

Rodent has a feature allowing to reduce its playing strength. For now, it is
still in the experimental phase and still undergoes tweaking and tuning, but 
basic infrastructure is already in place. It consists of three parts:

1. Weaker levels play slower. On starting a new iteration or changing a move
   Rodent "freezes" for a moment, depending on its Elo command.
   
2. Weaker levels add random value to evaluation score.

3. Weaker levels are more likely to blunder, because they forget 
   about calculating consequences of certain moves.

As You can probably guess, tuning the third element of the algorithm presents
the hardest challenge. Engine forgetting about certain moves (like captures
by a pawn or (re)captures of recently moved piece) would look unnatural. 
Engine forgetting about escapes from check would happily sack material for 
false checkmates. Engine not missing tactical moves would not be weak enough.

IX. INI FILE

rodent.ini file contains the following information

- panel type, determined by string "user nomal" or "user power"
- list of available strength levels
- list of available playing styles
- list of available opening books

Normal user panel lists available levels, strength and personalities, registered
in the ini file. Power user obtains the ability to manually set internal engine 
parameters via UCI options.

X. CONVENTIONS AND STRUCTURAL DECISIONS

Naming conventions:

- an asterisk before a comment announcing a code segment
  means that this part of code is undergoing modification 
  and testing
- variable names start with small letters
- function names start with captital letters
- variables treated as flags have "f" prefix, i.e. f_in_check
- bitboard variables have "bb" prefix, i.e. bbControl 
  in piece evaluators
- struct definitions have "s" prefix,  i.e. sEvaluator

Classes:

- sParser       - UCI parser  
- sTransTable   - transposition table (no functional change vs Sungorus)
- sHistory      - global history and killer tables
- sGenCache     - caching generated bitboards for minimal speedup
- sTimer        - setting and enforcing time controls
- sSelector     - picks moves in an ordered fashion
- sSearcher     - implements classical alpha-beta pvs search algorithm
- sEvaluator    - implements valuation function
- sData         - globally accessible configurable data; in the target version, 
                  if someone wants to concentrate on just one facet 
                  of a program, he will have to know just sData 
                  and the relevant class.

Structural decisions:

- we use the same square ordering as Crafty and Stockfish (A1 = 0)
- we DO NOT split Search function into pv and non-pv, 
  since resulting complexity isn't worth small speed gain.

changelog.txt
CHANGES AFTER 2013-01-23 recorded on GitHub, https://github.com/nescitus/rodent_code/commits/

2013-01-20: less cluttered display of search lines in console mode

2013-01-19: color-independent EvalPatterns() (currently only code for a knight blocking c-pawn moved there)

2013-01-17

- iid at cut nodes failed very narrowly (47,2 %)
- removed useFutility and useLmr variables, as they are practically always on and user can
  switch them off by setting lmrBase or futilityDepth

2013-01-16

- fixed silly sign bug in go nodes command
- internal iterative deepening at cut nodes passed self-play test and awaits for a gauntlet

2013-01-15: node type detection based on Fruit

2013-01-13

- shorter and marginally faster code for endgame knowledge
- bugfix (drawn KBPKB was not evaluated as 0 due to scaling opposite bishops first)
- bugfix in bench (history has not been cleared)

2013-01-09

- hanging pieces: 47,3% / 2249 games) 
- bench command implemented

2013-01-07

- failed attempts at tuning (at least several parameters' ranges are narrowed down) 
- FAILED with not evaluating the least advanced of double passed pawns
- FAILED applying candidate passer eval to all non-weak pawns
- moved perft rotines to a separate file
- removed book verification code

2013-01-06

- piece pairs tuning at fixed nodes DOESN'T HELP AT LONGER TIME CONTROLS
- variables "nodes" and "rootDepth" moved to sSearch class

2013-01-05

- FAILED with attacks on loose enemy pieces (44,5%/ 2300 games), even though it wins in self-test
- perft utility

2013-01-04

- FAILED attempt to implement piece lists (no speed gain and random effect on move ordering 
  is detrimental)

2013-01-03

- FAILED with pawn pst emphasizing center (45,5% / 2200 games)
- implemented "go movetime" and "go nodes" commands

2013-01-02: FAILED with simultaneous pawn pst change and attackable knight eval (44,2% / 2100 games)

2012-12-xx: FAILED with razoring a la DiscoCheck (perhaps better qs would be needed)

2012-12-30: aspiration window correction: 46,5% / 2000 games

2012-12-25: Eval pruning similar to that of DiscoCheck replaces Static null move: 46,2% / 2440 games

2012-12-21

- two bugfixes in material evaluation (but score not changed)
- restructuring material evaluation, which so far mixed two things; 
  now we have EvalMaterial() which does all the material adjustements 
  and CheckmateHelper() for pushing enemy king towards the rim and KBNK stuff

2012-12-20: confirmed that midgame eval of candidate passers is useful

2012-12-19: first attempt at candidate passer eval already gains a little: 46,0% / 1740 games

2012-12-18

- simplifying some code

2012-12-17

- failed with using History.MoveIsBad() as additional sorting criterion
- found a bug in pawn center evaluation
- coded pawn phalanx eval, but together with bugfix score drops to 45,3% / 2540 games

2012-12-16

- testing two null move alternatives ( different new depth formula saved in comments
  and reducing more when difference between static eva and beta is big)
  no improvement and no weakening from either of these formulas, so we stay conservative

2012-12-15: null move fix 45,7% / 2330 games

2012-12-14

- returning to work on Rodent
- overnight gauntlet 15s/game: 45,3% / 2340 games
- shrinking code initializing bad bishop mask (no functional change)

2012-10-09

- speedup due to different updating scheme of reversible moves counter (thx Kestutis Gasaitis!)
- failed with Kestitis' suggestion that reduction of null move verification search is too big  

2012-10-07: fixed sign bug in "checkmate help" in eval.material, added code for KBNK mate

2012-10-03: reporting nps every now and then in verbose mode

2012-10-02: separating strength and style settings

2012-09-24: faster function for loading just a move fom TT used in internal iterative deepening (thx Kestutis)

2012-09-21: pawn shelter halved on 4 central files (new gauntlet: 45,4%, 2172 games)

2012-09-14: changing tt date of used entries (thx Kestutis)

2012-09-12

- FIXED node discrepancy between debug and release version, but score fell to 33,8%
- FAILED to get speed increase from saving pst scores

2012-09-10

- small speedup due to different updating scheme of remaining moves
- separating sManipulator class for doing/undoing moves

2012-09-07: refined score stays

2012-09-05: determined that full eval for null move is better

2012-09-03: FAILED with two set of values for candidate passers

2012-09-02: FAILED with pawn phalanx code, should try again with different values

2012-08-28: reimplementing book loading and personality loading

2012-08-27: detection of missing book moves separated into a function and applied also in out of book positions

2012-08-25

- FAILED with king danger table
- opening book detects alternative moves leading to book positions (on finding a move only)

2012-08-23: FAILED change location of a line keeping history score within bounds

2012-08-22: PASSED narrowly with using hash information for sorting qs moves (thx Kestutis Gasaitis!)

2012-08-21: FAILED with null move verification (but added Kestutis' ply bugfix)

2012-08-18: FIXED side to move error that switched off black trapped bishop value

2012-08-17

- modified value returned by static null move /9thx Kestutis!)
- book utility converting e2e4e7e5g1f3b8c6 into e2e4 e7e5 g1f3 b8c6

16-08-2012

- no TT save of a null move result if we already have a hash move (thx Kestutis Gasaitis!)
- moved IID code after null move (doesn't matter right now, but improves future options, thx Kestutis)

2012-08-14

- FAILED with Fruit values for pawn storm
- re-implemented learning code

2012-08-07

- rewriting delta pruning code - result is the same, but code is more tunable
- FAILED with no TT in QS: 27,3%
- NO CHANGE, FAILED no delta pruning in PV QS nodes
- FAILED adding pawn delta = 100: 27,6% (is it a NPS thing?)
- FIXED sign error in Eval::Normalize() (thx Kestutis!)

2012-08-06

- returning to the highest-rated version 0.13 due to steady rating decline 

data.c
/*
  Rodent, a UCI chess playing engine derived from Sungorus 1.4
  Copyright (C) 2009-2011 Pablo Vazquez (Sungorus author)
  Copyright (C) 2011-2013 Pawel Koziol

  Rodent is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published 
  by the Free Software Foundation, either version 3 of the License, 
  or (at your option) any later version.

  Rodent is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty 
  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <math.h>
#include "data.h"
#include "bitboard/bitboard.h"
#include "rodent.h"

U64 bbLineMask[4][64];
U64 attacks[4][64][64];
U64 bbPawnAttacks[2][64];
U64 bbKnightAttacks[64];
U64 bbKingAttacks[64];
U64 bbKingZone[2][64];
U64 bbPassedMask[2][64];
U64 bbAdjacentMask[8];
U64 bbBadBishopMasks[2][64];
int castleMask[64];

// used for a mathematical trick in FirstOne
const int bitTable[64] = {
   0,  1,  2,  7,  3, 13,  8, 19,
   4, 25, 14, 28,  9, 34, 20, 40,
   5, 17, 26, 38, 15, 46, 29, 48,
  10, 31, 35, 54, 21, 50, 41, 57,
  63,  6, 12, 18, 24, 27, 33, 39,
  16, 37, 45, 47, 30, 53, 49, 56,
  62, 11, 23, 32, 36, 44, 52, 55,
  61, 22, 43, 51, 60, 42, 59, 58
};

U64 zobPiece[12][64];
U64 zobCastle[16];
U64 zobEp[8];
int pondering;
char ponder_str[6];

// mobility values
static const int n_mob_mg[28] = { -16-4, -12-2, -8,  0,  4,  8,  12, 16, 16};
static const int n_mob_eg[28] = { -16-4, -12-2, -8,  0,  4,  8,  12, 16, 16};
static const int b_mob_mg[28] = { -16-4, -12-2, -8, -4, -2,  0,  2,  4,  6,  8, 10, 12, 14, 14, 16                                 };
static const int b_mob_eg[28] = { -16-4, -12-2, -8, -4, -2,  0,  2,  4,  6,  8, 10, 12, 14, 14, 16                                 };
static const int r_mob_mg[28] = {  -6-4,  -4-2, -2,  0,  2,  4,  6,  8,  9, 10, 11, 12, 13, 14, 15                                 };
static const int r_mob_eg[28] = { -12-4,  -8-2, -4,  0,  4,  8, 12, 16, 18, 20, 22, 24, 26, 28, 30                                 };
static const int q_mob_mg[28] = {  -6-4,  -4-2, -2,  0,  0,  1,  1,  1,  2,  2,  2,  3,  3,  3,  3,  4,  4,  4,  4,  5,  5,  5,  5 };
static const int q_mob_eg[28] = { -12-4,  -8-2, -4, -2,  0,  1,  2,  3,  4,  5,  6,  6,  7,  7,  8,  8,  8,  9,  9,  9, 10, 10, 10 };


void sData::InitBadBishop(void)
{
	 // first clear bad bishop masks
	 for (int side = 0; side < 2; side++) {
	    for (int i = 0; i < 64; i++) {  
			 bbBadBishopMasks [side][i] = 0ULL;
		 }
	 }

	 SetBadBishopMask(F1, E2);
	 SetBadBishopMask(C1, D2);
	 SetBadBishopMask(A2, B3); SetBadBishopMask(A2, C4);
	 SetBadBishopMask(B2, C3);
	 SetBadBishopMask(C2, D3); SetBadBishopMask(C2, E4);
	 SetBadBishopMask(D2, E3);
	 SetBadBishopMask(E2, D3);
	 SetBadBishopMask(F2, E3); SetBadBishopMask(F2, D4);
	 SetBadBishopMask(G2, F3);
	 SetBadBishopMask(H2, G3); SetBadBishopMask(H2, F4);
	 SetBadBishopMask(A3, B4); SetBadBishopMask(A3, C5);
	 SetBadBishopMask(B3, C4); SetBadBishopMask(B3, D5);
	 SetBadBishopMask(C3, D4);
	 SetBadBishopMask(D3, E4);
	 SetBadBishopMask(E3, D4);
	 SetBadBishopMask(F3, E4);
	 SetBadBishopMask(G3, F4); SetBadBishopMask(G3, E5);
	 SetBadBishopMask(H3, G4); SetBadBishopMask(H3, F5);
	 SetBadBishopMask(C4, D5);
	 SetBadBishopMask(F4, E5);
}

void sData::SetBadBishopMask(int bishSq, int pawnSq)
{
     bbBadBishopMasks[WHITE][bishSq] ^= SqBb(pawnSq);
	 bbBadBishopMasks[BLACK][ REL_SQ(bishSq,BLACK) ] ^= RelSqBb(pawnSq, BLACK);
}

void sData::InitSinglePiece(int pc, int mat, int del, int phase, int att) 
{
     matValue[pc]      = mat;   
	 deltaValue[pc]    = del;
	 phaseValue[pc]    = phase;     
	 attMultiplier[pc] = att;      
}

void sData::InitMobBonus(void) 
{
	 for (int i = 0; i < 28; i++) {
		 mobBonusMg[N][i] = n_mob_mg[i];
		 mobBonusEg[N][i] = n_mob_eg[i];
		 mobBonusMg[B][i] = b_mob_mg[i];
		 mobBonusEg[B][i] = b_mob_eg[i];
		 mobBonusMg[R][i] = r_mob_mg[i];
		 mobBonusEg[R][i] = r_mob_eg[i];
		 mobBonusMg[Q][i] = q_mob_mg[i];
		 mobBonusEg[Q][i] = q_mob_eg[i];
	 }
}

void sData::InitAttackBonus(void) 
{
    int startCurve = 10; // 7 is worse
	int logMult    = 20; // 20 is OK, raising doesn't seem to help

	 for (int i=0; i < 256; i++)
     attackBonus[i] = i / 3;

	 for (int i=0; i < 256; i++) {
		attackBonus[i+startCurve] = logMult * log((double) i) + i / 2;
		attackBonus[startCurve] = startCurve / 2;
		attackBonus[startCurve+1] = startCurve / 2+2;
	    //printf("%2d %d\n", i, attackBonus[i] );
  }
}

void sData::InitCastleMask(void) 
{
  for (int i = 0; i < 64; i++) castleMask[i] = 15;
  castleMask[A1] = 13;
  castleMask[E1] = 12;
  castleMask[H1] = 14;
  castleMask[A8] = 7;
  castleMask[E8] = 3;
  castleMask[H8] = 11;
}

void sData::InitSearchData(void) 
{
     // History of moveIsLate:lmrStep tuning:
	 // 10:5 best, 8:4 best, 8:5 fails, 6:4 best, 5:4 fails)

     aspiration       = 30;
	 useNull          = 1;
	 minimalNullDepth = ONE_PLY;
	 minimalLmrDepth  = 2 * ONE_PLY;
	 moveIsLate       = 6;
	 lmrStep          = 4; 
	 lmrHistLimit     = 60;      // 70 is better in very fast games (10s per game)
	 futilityBase     = 100;
	 futilityStep     = 20;
	 futilityDepth    = 4;
	 useDeltaPruning  = 1;
	 deltaMargin      = 150;
	 goodCaptMargin   = 30;      // BxN is OK, RxB isn't
	 lazyMargin       = 180;
	 verbose          = 0;       // no additional display
	 elo              = MAX_ELO; // no weakening
	 isAnalyzing      = 0;
	 useBook          = 1;
	 useWeakening     = 0;
	 useLearning      = 0;
}

void sData::InitDistanceBonus(void) 
{
    for (int i = 0; i < 64; ++i) {
         for (int j = 0; j < 64; ++j) {
         distance[i][j] = 14 - ( Abs( Rank(i) - Rank(j) ) + Abs( File(i) - File(j) ) );
         }
     }
}

void sData::InitMaterialValues(void) 
{
	//             name,  val, del, pha, att
	 InitSinglePiece( P,  0,   150,  0,   0 );  // pawn material is evaluated in eval.c
	 InitSinglePiece( N,  325, 475,  1,   2 );
	 InitSinglePiece( B,  335, 485,  1,   2 );
	 InitSinglePiece( R,  500, 650,  2,   4 );  // 510 is worse
	 InitSinglePiece( Q,  975, 1125, 4,   7 );
	 InitSinglePiece(NO_TP, 0, 150,  0,   0 );
};

void sData::InitEvalVars(void) 
{
	 ownMobility     = 110;
	 oppMobility     = 110;
	 ownAttack       = 100; 
	 oppAttack       = 100; 
	 bishopPair      =  50; 
	 doubledPawn[MG] = -20;
     doubledPawn[EG] = -10;
	 rookOpenMg      = 10;
     rookOpenEg      = 10;
	 rookSemiOpenMg  = 5;
	 rookSemiOpenEg  = 5;
	 rookOpenAttack  = 10;
	 rookSemiOpenAttack = 5;
	 rookSeventhMg   = 20;
	 rookSeventhEg   = 20;
	 pawnIsolatedOnOpen = -15;
	 pawnBackwardOnOpen = -15;
}

// used at the beginning of search to set scaling factors for eval components
void sData::InitAsymmetric(int side) 
{
     mobSidePercentage[side]      = ownMobility;
     mobSidePercentage[Opp(side)] = oppMobility;
     attSidePercentage[side]      = ownAttack;
     attSidePercentage[Opp(side)] = oppAttack;
}